File: ranchsupply-backend/ranchsupply-backend/src/main/java/org/launchcode/ranchsupply/service/CategoryService.java
package org.launchcode.ranchsupply.service;

import org.launchcode.ranchsupply.exception.ResourceNotFoundException;
import org.launchcode.ranchsupply.model.Category;
import org.launchcode.ranchsupply.model.dto.CategoryDto;
import org.launchcode.ranchsupply.repository.CategoryRepository;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
public class CategoryService {
    @Autowired
    private CategoryRepository categoryRepository;

    @Autowired
    private ModelMapper modelMapper;

    //to add Category
    public CategoryDto addCategory(CategoryDto categoryDto){
        Category category = modelMapper.map(categoryDto, Category.class);
        Category savedCategory = categoryRepository.save(category);
        return modelMapper.map(savedCategory,CategoryDto.class);
    }

    public CategoryDto getCategoryById(Long categoryId) {
        Category category = categoryRepository.findById(categoryId)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found with ID: " + categoryId));
        return modelMapper.map(category, CategoryDto.class);
    }

    // Get All the categories
    public List<CategoryDto> getAllCategories(){
        List<Category> categories = categoryRepository.findAll();
        return categories.stream().map((category) -> modelMapper.map(category,CategoryDto.class))
                .collect(Collectors.toList());
    }

    public CategoryDto updateCategory(Long categoryId, CategoryDto categoryDto) {
        Category existingCategory = categoryRepository.findById(categoryId)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found"));
        modelMapper.map(categoryDto, existingCategory);
        Category updatedCategory = categoryRepository.save(existingCategory);
        return modelMapper.map(updatedCategory, CategoryDto.class);
    }

    public void deleteCategory(Long categoryId) {
        Category category = categoryRepository.findById(categoryId)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found"));
        categoryRepository.delete(category);
    }
}

File: ranchsupply-frontend/src/services/CategoryService.js
import axios from "axios";

const REST_API_BASE_URL = "http://localhost:8080/categories";

// Get all categories
export const getAllCategories = async () => {
  const response = await axios.get(REST_API_BASE_URL);
  return response.data;
};

// Get a category by ID
export const getCategoryById = async (categoryId) => {
  const response = await axios.get(`${REST_API_BASE_URL}/${categoryId}`);
  return response.data;
};

// Add a new category
export const addCategory = async (category) => {
  const response = await axios.post(REST_API_BASE_URL, category, {
    headers: {
      'Content-Type': 'application/json'
    }
  });
  return response.data;
};

// Update an existing category
export const updateCategory = async (categoryId, category) => {
  const response = await axios.put(`${REST_API_BASE_URL}/${categoryId}`, category, {
    headers: {
      'Content-Type': 'application/json'
    }
  });
  return response.data;
};

// Delete a category by ID
export const deleteCategory = async (categoryId) => {
  await axios.delete(`${REST_API_BASE_URL}/${categoryId}`);
};
